components/gfx/paint_context.rs

```diff
@@ -355,6 +368,102 @@ impl<'a> PaintContext<'a> {
         self.draw_target.push_clip(&path_builder.finish());
     }
 
+    fn solve_quadratic(a: f32, b: f32, c: f32) -> (Option<f32>, Option<f32>) {
+        let descriminant = b * b - 4. * a * c;
+        if descriminant < 0. {
+            return (None, None);
+        }
+        let x1 = (-b + descriminant.sqrt())/(2. * a);
+        let x2 = (-b - descriminant.sqrt())/(2. * a);
+        if descriminant == 0. {
+            return (Some(x1), None);
+        }
+        return (Some(x1), Some(x2));
+    }
+
+    fn intersect_ellipse_line(e: Ellipse, line: Line) -> (Option<Point2D<f32>>, Option<Point2D<f32>>)  {
+        if line.end.x - line.start.x <= f32::EPSILON {
+            panic!("Error line segment end.x > start.x!!");
+        }
+        // shift the origin to center of the ellipse.
+        let l = Line { start: Point2D { x: line.start.x - e.origin.x,
+                                        y: line.start.y - e.origin.y},
+                       end: Point2D { x: line.end.x - e.origin.x,
+                                      y: line.end.y - e.origin.y}};
+
+        let a = (l.end.y - l.start.y)/(l.end.x - l.start.x);
+        let b = l.start.y - (a * l.start.x);
+        // given the equation of a line,
+        // y = a * x + b,
+        // and the equation of an ellipse,
+        // x^2/w^2 + y^2/h^2 = 1,
+        // substitute y = a * x + b, giving
+        // x^2/w^2 + (a^2x^2 + 2abx + b^2)/h^2 = 1
+        // then simplify to
+        // (h^2 + w^2a^2)x^2 + 2abw^2x + (b^2w^2 - w^2h^2) = 0
+        // finally solve for w using the quadratic equation.
+        let w = e.width;
+        let h = e.height;
+        let quad_a = h * h + w * w * a * a;
+        let quad_b = 2. * a * b * w * w;
+        let quad_c = b * b * w * w - w * w * h * h;
+        let intersections = PaintContext::solve_quadratic(quad_a, quad_b, quad_c);
+        match intersections {
+            (Some(x0), Some(x1)) => {
+                let mut p0 = Point2D {x: x0 + e.origin.x, y: a * x0 + b + e.origin.y};
+                let mut p1 = Point2D {x: x1 + e.origin.x, y: a * x1 + b + e.origin.y};
+                if x0 > x1 {
+                    let tmp = p0;
+                    p0 = p1;
+                    p1 = tmp;
+                }
+                (Some(p0), Some(p1))
+            },
+            (Some(x0), None) => {
+                let p = Point2D {x: x0 + e.origin.x, y: a * x0 + b + e.origin.y};
+                (Some(p), None)
+            },
+            (None, Some(x1)) => {
+                let p = Point2D {x: x1 + e.origin.x, y: a * x1 + b + e.origin.y};
+                (Some(p), None)
+            },
+            (None, None) => (None, None),
+        }
+    }
+
+    fn ellipse_line_intersection_angles(e: Ellipse, l: Line)
+                                        -> (Option<(Point2D<f32>, f32)>, Option<(Point2D<f32>, f32)>) {
```

Comment:
Can you add a comment explaining what this function does? I'm guessing that the points are the intersection points and the other `f32`s are angles of some kind, but I don't know which angles they are and I'd prefer to have a comment here rather than forcing everyone who reads this code to draw it out :)